### PR TITLE
[CORE-766] PreInstall data using Google Referrer api

### PR DESCRIPTION
**Pull request by msiddiq-branch**
## Reference
[CORE-766] PreInstall data using Google Referrer API

## Description 
For Preinstall campaigns client had to manually add a condition in the APK to set the 3p value and campaign information based on the value read from Google referrer API. Since we are already capturing this information at the SDK level, have added method to set PreInstall data at SDK itself.  It would set the data only when utmSource=play-auto-installs. utm_campaign is set to PreInstallCampaign and utm_medium is set to PreinstallPartner.  

https://branch.atlassian.net/browse/CORE-766

**Priority on different PreInstall method**
1. Manual setting for PreInstall value
2. OS level 
3. Google Play referrer

## Testing Instructions
**Usecase 1**:
Step 1: Click on the below google playstore link which has utm param
https://play.google.com/store/apps/details?id=io.branch.branchandroiddemo&referrer=utm_source%3Dplay-auto-installs%26utm_medium%3Da_xiaomipai%26utm_campaign%3DMI
Step 2: Make sure PreInstall method 1 & 2 are not set
Step 3: Install the app from Android Studio 
Step 4: Check live view for 3p and campaign names being captured for INSTALL/OPEN 
**Expected result**:
Should see the event(INSTALL/OPEN) being attributed to value passed utm_medium and utm_campaign which is set in the above play store link

**Usecase 2**:
Step 1: Click on the below google playstore link which has utm param
https://play.google.com/store/apps/details?id=io.branch.branchandroiddemo&referrer=utm_source%3Dplay-auto-installs%26utm_medium%3Da_xiaomipai%26utm_campaign%3DMI
Step 2: Make sure PreInstall method 1 or 2 is set
Step 3: Install the app from Android Studio 
Step 4: Check live view for 3p and campaign names being captured for INSTALL/OPEN event
**Expected result**:
Should see the event(INSTALL/OPEN) being attributed to the value passed under PreInstall method 1 or 2

## Risk Assessment [`HIGH` || `MEDIUM` || `LOW`]

- [ ] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- [ ] JIRA Ticket is referenced in PR title.
- Correctness & Style
    - [ ] Conforms to [AOSP Style Guides](https://source.android.com/setup/contribute/code-style)
    - [ ] Mission critical pieces are documented in code and out of code as needed.
- [ ] Unit Tests reviewed and test issue sufficiently.
- [ ] Functionality was reviewed in QA independently by another engineer on the team.

[CORE-766]: https://branch.atlassian.net/browse/CORE-766

### CODE REVIEW

File: Branch-SDK/src/main/java/io/branch/referral/GooglePlayStoreAttribution.java [hunk 1]
```diff
@@ -143,7 +143,13 @@ private static void processReferrerInfo(Context context, String rawReferrerStrin
                     prefHelper.setGoogleSearchInstallIdentifier(referrerMap.get(Defines.Jsonkey.GoogleSearchInstallReferrer.getKey()));
                     prefHelper.setGooglePlayReferrer(rawReferrerString);
                 }
-                
+
+                if(referrerMap.containsValue(Defines.Jsonkey.PlayAutoInstalls.getKey())) {
+
+                    prefHelper.setGooglePlayReferrer(rawReferrerString);
+                    BranchPreinstall.setBranchPreInstallGoogleReferrer(context,referrerMap);
```
Comment (code-crusher) [inline Branch-SDK/src/main/java/io/branch/referral/GooglePlayStoreAttribution.java]:
```suggestion
                    BranchPreinstall.setBranchPreInstallGoogleReferrer(context, referrerMap);
```

File: Branch-SDK/src/main/java/io/branch/referral/GooglePlayStoreAttribution.java [hunk 2]
```diff
@@ -143,7 +143,13 @@ private static void processReferrerInfo(Context context, String rawReferrerStrin
                     prefHelper.setGoogleSearchInstallIdentifier(referrerMap.get(Defines.Jsonkey.GoogleSearchInstallReferrer.getKey()));
                     prefHelper.setGooglePlayReferrer(rawReferrerString);
                 }
-                
+
+                if(referrerMap.containsValue(Defines.Jsonkey.PlayAutoInstalls.getKey())) {
+
```
Comment (code-crusher) [inline Branch-SDK/src/main/java/io/branch/referral/GooglePlayStoreAttribution.java]:
```suggestion
```

File: Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java [hunk 3]
```diff
@@ -114,4 +115,26 @@ public static void getBranchFileContent(JSONObject branchFileContentJson,
             }
         }
     }
+
+    public static void setBranchPreInstallGoogleReferrer(Context context, HashMap<String, String> referrerMap){
+
```
Comment (code-crusher) [inline Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java]:
```suggestion
```

File: Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java [hunk 4]
```diff
@@ -114,4 +115,26 @@ public static void getBranchFileContent(JSONObject branchFileContentJson,
...
+        Branch branchInstance = Branch.getInstance();
+        PrefHelper prefHelper = PrefHelper.getInstance(context);
+
+        //Set PreInstallData from GoogleReferrer api
```
Comment (code-crusher) [inline Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java]:
```suggestion
        // Set PreInstallData from GoogleReferrer api
```

File: Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java [hunk 5]
```diff
@@ -114,4 +115,26 @@ public static void getBranchFileContent(JSONObject branchFileContentJson,
...
+        // only if PreInstallMetaData has not been updated by either of the methods(Manual setting or OS level)
+        if(!(prefHelper.getInstallMetaData(PreinstallKey.partner.getKey())!=null || PrefHelper.getInstance(context).getInstallMetaData(PreinstallKey.campaign.getKey())!=null)){
+
```
Comment (code-crusher) [inline Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java]:
```suggestion
```

File: Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java [hunk 6]
```diff
@@ -114,4 +115,26 @@ public static void getBranchFileContent(JSONObject branchFileContentJson,
...
+            if(referrerMap.get(Defines.Jsonkey.UTMCampaign.getKey())!=null){
+                branchInstance.setPreinstallCampaign(referrerMap.get(Defines.Jsonkey.UTMCampaign.getKey()));
+            }
+
```
Comment (code-crusher) [inline Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java]:
```suggestion
```

File: Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java [hunk 7]
```diff
@@ -114,4 +115,26 @@ public static void getBranchFileContent(JSONObject branchFileContentJson,
...
     }
+
+    public static void setBranchPreInstallGoogleReferrer(Context context, HashMap<String, String> referrerMap){
+
+        Branch branchInstance = Branch.getInstance();
+        PrefHelper prefHelper = PrefHelper.getInstance(context);
+
+        //Set PreInstallData from GoogleReferrer api
+        // only if PreInstallMetaData has not been updated by either of the methods(Manual setting or OS level)
+        if(!(prefHelper.getInstallMetaData(PreinstallKey.partner.getKey())!=null || PrefHelper.getInstance(context).getInstallMetaData(PreinstallKey.campaign.getKey())!=null)){
```
Comment (code-crusher) [inline Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java]: Let's use `TextUtils.isEmpty` to check for empty and null at the same time.

File: Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java [hunk 8]
```diff
@@ -114,4 +115,26 @@ public static void getBranchFileContent(JSONObject branchFileContentJson,
...
+
+            if(referrerMap.get(Defines.Jsonkey.UTMCampaign.getKey())!=null){
```
Comment (code-crusher) [inline Branch-SDK/src/main/java/io/branch/referral/BranchPreinstall.java]: Same here `TextUtils.isEmpty`

**Comment by code-crusher** [comment]
@msiddiq-branch were you able to test this in some way? I do not want the install event to get fired before this is set.

The setter method originally is called in the Application class, so it gets set before branch sens install request. Not sure about when we get the referrer API response. We need to validate this.

**Comment by msiddiq-branch** [comment]
> @msiddiq-branch were you able to test this in some way? I do not want the install event to get fired before this is set.
> 
> The setter method originally is called in the Application class, so it gets set before branch send install request. Not sure about when we get the referrer API response. We need to validate this.

@code-crusher I tested looking at log data of INSTALL API when the app is installed. Further, the condition of Install event being triggered before the value is set has been taken care as preinstall values are set at the time when installBeginTimeStamp is been set which I believe this has sent successfully every time install event is fired. 

{
...
...
   "latest_install_time":1585243124875,
   "latest_update_time":1585243124875,
   "first_install_time":1585243124875,
   "cd":{
      "mv":"-1",
      "pn":"io.branch.branchandroiddemo"
   },
   "clicked_referrer_ts":1585243108,
   **"preinstall_campaign":"MI",
   "preinstall_partner":"a_xiaomipai",**
   "install_referrer_extras":"utm_source=play-auto-installs&utm_medium=a_xiaomipai&utm_campaign=MI",
 ...
}

**Comment by bklastaitis-branch** [comment]
@msiddiq-branch @code-crusher I was confused when reading the Jira ticket, because it said `utmMedium=preload` is fixed and set by Google but the testing url contained `utm_medium=a_xiaomipai`.. Then I saw that it was discussed with Prashant too, but I couldn't understand it fully, especially what the "3p value" means, so  I gave up.. I would love it if someone explained it to me, but otherwise the code looks good and doesn't break anything so I'm merging it.

FYI, since this is an important client you may or may not want to run a more robust integration test by making an "Internal release" of a custom build of BMF with a custom build of the SDK in it and a real air-preload campaign URL.

**Comment by msiddiq-branch** [comment]
@bklastaitis-branch Basically customer request was to capture campaign and partner information for preinstall from google play referrer api. 

Solution:
When Google play referrer has utm_source=play-auto-installs we will mark it as preinstall campaign and set the campaign name and 'Branch $3p Parameter Value' to preinstall_campaign and preinstall_partner respectively.

**Pull request merged**